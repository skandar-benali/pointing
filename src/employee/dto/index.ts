export * from './create-employee.dto';
export * from './find-all-employee.dto';
