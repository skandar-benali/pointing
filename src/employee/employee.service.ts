import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { Employee } from './entities';
import { PrismaService } from '../prisma.service';

@Injectable()
export class EmployeeService {
  constructor(private prisma: PrismaService) {}

  async create(employee: Prisma.EmployeeCreateInput): Promise<Employee> {
    return this.prisma.employee.create({
      data: employee,
    });
  }

  async findMany(
    filter: Prisma.EmployeeWhereInput = {},
    { page, limit }: { page: number; limit: number },
  ): Promise<Employee[]> {
    return this.prisma.employee.findMany({
      where: filter,
      skip: (page - 1) * limit,
      take: limit,
    });
  }

  async findOne(id: number): Promise<Employee | null> {
    return this.prisma.employee.findUnique({
      where: { id },
    });
  }

  async delete(id: number): Promise<Employee> {
    return await this.prisma.employee.delete({
      where: { id },
    });
  }
}
