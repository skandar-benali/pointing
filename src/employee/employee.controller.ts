import {
  Controller,
  Get,
  Post,
  Body,
  Param,
  Delete,
  Query,
} from '@nestjs/common';
import { CreateEmployeeDto, findAllEmployeeDto } from './dto';
import { EmployeeCore } from './employee.core';

@Controller('employee')
export class EmployeeController {
  constructor(private readonly employeeCore: EmployeeCore) {}

  @Post()
  create(@Body() createEmployeeDto: CreateEmployeeDto) {
    return this.employeeCore.add(createEmployeeDto);
  }

  @Get()
  findAll(@Query() query: findAllEmployeeDto) {
    return this.employeeCore.list({
      startDate: query?.startDate,
      endDate: query?.startDate,
      page: query?.page ? Number(query?.page) : 1,
      limit: query?.limit ? Number(query?.limit) : 10,
    });
  }

  @Get(':id')
  findOne(@Param('id') id: number) {
    return this.employeeCore.findOne(Number(id));
  }

  @Delete(':id')
  remove(@Param('id') id: number) {
    return this.employeeCore.remove(Number(id));
  }
}
