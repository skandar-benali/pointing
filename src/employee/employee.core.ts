import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { Employee } from './entities';
import { EmployeeService } from './employee.service';
import { CreateEmployeeDto, findAllEmployeeDto } from './dto';

@Injectable()
export class EmployeeCore {
  constructor(private employeeService: EmployeeService) {}

  async add(employee: CreateEmployeeDto): Promise<Employee> {
    return this.employeeService.create(employee);
  }

  async list(filter: findAllEmployeeDto): Promise<Employee[]> {
    return this.employeeService.findMany(
      {
        ...(filter?.startDate && { createdAt: { gte: filter?.startDate } }),
        ...(filter?.endDate && { createdAt: { gte: filter?.startDate } }),
      },
      {
        page: filter?.page,
        limit: filter.limit,
      },
    );
  }

  async findOne(id: number): Promise<Employee | null> {
    const employee = await this.employeeService.findOne(id);
    if (!employee) throw new HttpException('Not Found', HttpStatus.NOT_FOUND);
    return employee;
  }

  async remove(id: number): Promise<Employee> {
    const employee = await this.employeeService.findOne(id);
    if (!employee) throw new HttpException('Not Found', HttpStatus.NOT_FOUND);
    return await this.employeeService.delete(id);
  }
}
