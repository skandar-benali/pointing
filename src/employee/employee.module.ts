import { Module } from '@nestjs/common';
import { EmployeeService } from './employee.service';
import { EmployeeController } from './employee.controller';
import { PrismaService } from '../prisma.service';
import { EmployeeCore } from './employee.core';

@Module({
  controllers: [EmployeeController],
  providers: [EmployeeService, PrismaService, EmployeeCore],
})
export class EmployeeModule {}
