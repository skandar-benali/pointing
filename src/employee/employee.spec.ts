import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { EmployeeModule } from './employee.module';
import { INestApplication } from '@nestjs/common';
import { EmployeeService } from './employee.service';

describe('/employee POST create employee', () => {
  let app: INestApplication;
  const user = {
    firstName: 'tester',
    name: 'tester',
    department: 1,
  };
  const employeeService = {
    create: () => ({
      ...user,
      id: 1,
      createdAt: '2021-12-16T10:32:04.710Z',
      updatedAt: '2021-12-16T10:32:04.710Z',
    }),
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [EmployeeModule],
    })
      .overrideProvider(EmployeeService)
      .useValue(employeeService)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  test('should return the created employee', async () => {
    // console.log(res.);

    return request(app.getHttpServer())
      .post('/employee')
      .send(user)
      .expect(201)
      .then((response) => {
        expect(response.body).toMatchSnapshot();
      });
  });

  afterAll(async () => {
    await app.close();
  });
});
