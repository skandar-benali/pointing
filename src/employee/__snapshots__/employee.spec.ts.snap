// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/employee POST create employee should return the created employee 1`] = `
Object {
  "createdAt": "2021-12-16T10:32:04.710Z",
  "department": 1,
  "firstName": "tester",
  "id": 1,
  "name": "tester",
  "updatedAt": "2021-12-16T10:32:04.710Z",
}
`;
